### PR TITLE
Add Grpc-specific Kryo serializers (fix #2963)

### LINKED ISSUE #2963: Add Kryo serializers to registry for GRPC exception types
some of them have inner references that aren't automatically serializable

sample CoderTest:
```scala
import io.grpc.StatusRuntimeException

it should "support GRPC exceptions" in {
  val tryFailure: Try[String] = Failure(new StatusRuntimeException(Status.OK))
  tryFailure coderShould roundtrip()
}
```
```
[info] -support GRPC exceptions *** FAILED *** (44 milliseconds)
[info]   java.lang.RuntimeException: Exception while trying to `encode` an instance of scala.util.Failure:  Can't encode field exception value io.grpc.StatusRuntimeException: OK
[info]   at com.spotify.scio.coders.RecordCoder.onErrorMsg(Coder.scala:267)
[info]   at com.spotify.scio.coders.RecordCoder.encode(Coder.scala:279)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:211)
[info]   at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:221)
[info]   at com.spotify.scio.coders.LazyCoder.encode(Coder.scala:183)
[info]   at com.spotify.scio.coders.DisjunctionCoder.encode(Coder.scala:126)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
[info]   ...
[info]   Cause: com.esotericsoftware.kryo.KryoException: Error during Java serialization.
[info]   at com.esotericsoftware.kryo.serializers.JavaSerializer.write(JavaSerializer.java:54)
[info]   at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:651)
[info]   at com.spotify.scio.coders.KryoAtomicCoder$$anonfun$encode$1.apply(KryoAtomicCoder.scala:147)
[info]   at com.spotify.scio.coders.KryoAtomicCoder$$anonfun$encode$1.apply(KryoAtomicCoder.scala:137)
[info]   at com.spotify.scio.coders.KryoAtomicCoder$.withKryoState(KryoAtomicCoder.scala:282)
[info]   at com.spotify.scio.coders.KryoAtomicCoder.encode(KryoAtomicCoder.scala:137)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
[info]   at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:211)
[info]   ...
[info]   Cause: java.io.NotSerializableException: io.grpc.Status
[info] 	- field (class "io.grpc.StatusRuntimeException", name: "status", type: "class io.grpc.Status")
[info] 	- root object (class "io.grpc.StatusRuntimeException", io.grpc.StatusRuntimeException: OK)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1182)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[info]   at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
[info]   at com.esotericsoftware.kryo.serializers.JavaSerializer.write(JavaSerializer.java:51)
[info]   at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:651)
[info]   at com.spotify.scio.coders.KryoAtomicCoder$$anonfun$encode$1.apply(KryoAtomicCoder.scala:147)
[info]   at com.spotify.scio.coders.KryoAtomicCoder$$anonfun$encode$1.apply(KryoAtomicCoder.scala:137)
```

### PR DESCRIPTION
**Pull request by ClaireMcGinty**
fix #2963

### CODE REVIEW

File: scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      )
+    )
+  }
+
+  it should "roundtrip with all nullable fields absent" in {
+    roundtrip(Failure(new StatusRuntimeException(Status.OK)))
+  }
+
+  private def roundtrip(t: Try[String]): Unit = {
+    val kryoBCoder = CoderMaterializer.beamWithDefault(Coder[Try[String]])
```
Comment (regadas) [inline scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: Something looks off here ... the usages above should have never compiled since they should be of type `Try[StatusRuntimeException]` 🤔

Comment (regadas) [reply scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: I guess the only reason is somehow the type inference is off and it's something like `Failure[Nothing]`

File: scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala [hunk 2]
```diff
@@ -122,7 +122,9 @@ final private class ScioKryoRegistrar extends IKryoRegistrar {
     k.forSubclass[Path](new JPathSerializer)
     k.forSubclass[ByteString](new ByteStringSerializer)
     k.forClass(new KVSerializer)
-    k.addDefaultSerializer(classOf[Throwable], new JavaSerializer())
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala:125]: why was this changed?

File: scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala [hunk 3]
```diff
@@ -122,7 +122,9 @@ final private class ScioKryoRegistrar extends IKryoRegistrar {
...
+    k.forClass[io.grpc.Status](new grpc.StatusSerializer())
+    k.javaForClass[Throwable]
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala:126]: `javaForClass` is not really the same thing as the line above

File: scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala (hunk 1 above)
Comment (ClaireMcGinty) [reply scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: Ah no, it's supposed to be `Try[String]`, it's kind of a shortcut mimicking a real use case with AsyncLookupDoFn where the return type is something like `Try[GrpcResponse]`, which is either a `Success(GrpcResponse)` or `Failure(StatusRuntimeException)`. So I'm testing that (a) the Kryo serializer works and (b) that it's being picked up in a `Coder[Try[T]]` instead of the default Kryo coder registered for `Throwable`

Comment (regadas) [reply scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: Ahh, right gotcha .. being `Try[String]` is confusing make it just `Try[T]` would be better.

Also, I think here we should be only testing `Coder[StatusRuntimeException]`. The semantic of picking a coder for a given `T` should be covered elsewhere.

Comment (ClaireMcGinty) [reply scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: 👍 

that's fair, I wrote the test like this because in most use cases we wouldn't have a static `Coder[StatusRuntimeException]` defined somewhere, the flow is like: AsyncLookupDoFn returns a `Try[GrpcResponse]` with a static `Coder[Try[GrpcResponse]]` --> result is determined to be a `Failure` so `FallbackCoder[Throwable]` is invoked --> Kryo deserializes the class name and correctly selects `StatusRuntimeExceptionSerializer` over `JavaSerializer` to deserialize the result. I can rework the test just to cover the last step :) (in a followup pr since this one is merged)

Comment (regadas) [reply scio-test/src/test/scala/com/spotify/scio/coders/instances/kryo/GrpcSerializersTest.scala:50]: Yup testing the last step is really what we need here 👍 

using `Coder[StatusRuntimeException]`, even if we don't foresee it's derivation, will just also make sure that we Fallback when we do.

However the thing that worries me more is the change on line L125. I would like to keep the default serializer otherwise we might have issues with subclasses.
